### ISSUE TITLE
Discord Threads support

### ISSUE THREAD

**Issue by alvierahman90** [matrix-discord/mx-puppet-discord#164]
Discord just launched [Threads](https://support.discord.com/hc/tr/articles/4403205878423-Threads). From the [Discord Developer Docs](https://discord.com/developers/docs/topics/threads):

> Threads can be thought of as temporary sub-channels inside an existing channel, to help better organize conversation in a busy channel.

There's no parallel to this in Matrix as far as I can tell so the only solution I can think of so far is to create a new room on the Matrix end... which is what Discord is trying to avoid on the Discord end but with double puppeting enabled you could always have the bot move archived threads to low priority and move it back if it gets un-archived, which makes clutter less of an issue.

You could also prefix messages in a thread with _Thread: thread name_ or something to that effect if you don't want to create new rooms.

Regardless of how it's implemented though (maybe make it configurable), I feel it's important this feature is bridged over when possible since if the servers you are part of make heavy use of Discord Threads, you will miss out on whole conversations.

**Comment by MRAAGH** [comment]
I think inviting to a new Matrix room is the way to go. And anything is better than nothing; currently I need to use Discord's app to be able to access threads

**Comment by yousefamar** [comment]
Threads in mx-puppet-slack are simply flattened into reply chains. I think this could be a good compromise as a quick first step? (although I know they're not quite the same thing, more like zulip.com rather than Slack)

In the long run, wouldn't nested Spaces be the more logical mapping? It's supposed to be possible to put Spaces in Spaces like a directory tree, though I don't think any clients currently have any good UI for that. The only weird thing might be to have to represent the parent channels as a Space-of-one unless Matrix Spaces can also behave like rooms at the top level (so "everything is a Space"). I do like that Discord threads are designed to be ephemeral or else this would be super hard to fit into Matrix, but it's still primarily going to be a UI/UX problem

**Comment by alvierahman90** [comment]
Flattened replies could work but I feel like that would result in a worse user experience on the matrix side since the whole aim of threads is to avoid having conversations be hidden by lots of messages in very busy servers. 

Nested spaces is a good idea I hadn't thought of that 👍

**Comment by MRAAGH** [comment]
Does this require something to be added to Matrix specification?

**Comment by yousefamar** [comment]
> To avoid spaces of one perhaps a space could be made automatically by the bot only once a thread has been made in a channel?

If Spaces work the way I think they do, I think this would be the best. So basically, the moment that a channel/room has threads, a new Space is created e.g. named after that room, then the "parent room" and the "thread room" are moved into that Space. I recall that moving rooms like this may be an expensive operation, but I'm not sure.

> Does this require something to be added to Matrix specification?

I don't think so (unless Matrix introduced the notion of Threads itself). The main problem is with the clients -- even Element at the moment can barely handle normal Spaces properly, so that's why I think it will have to be some kind of hack in the meantime (e.g. new rooms like you mentions) just so we have some way of seeing threads.

**Comment by sdfg2** [comment]
I'm not sure forcing a particular behaviour is the way forward, it would need to be optional.  I'm the kind of user who uses matrix and the bridges as a way of collating all my chat into one place, with a flat overview of all channels and DMs in a list.  My discord servers are created as spaces, which is handy for grouping if I need it, but otherwise the spaces are hidden in my client.  Flat list of all channels within all servers, sorted by activity, custom notifications where needed.

Don't get me wrong, I can see a benefit of nesting, but as has been stated, most current clients can't represent that well - they've essentially emulated the discord interface as it is.  Make it optional for people who don't mind the navigation problems currently, and hopefully someone comes out with a tree-based sidebar for representing the nesting for old-timers like me ;-)

**Comment by MRAAGH** [comment]
If we have multiple implementations, we could make it configurable in config.yaml, or even configurable per puppet ID through the bridge bot room. Although it may be better to just have one temporary solution until Matrix (and the clients) support threads in a way equivalent to Discord and Slack.

**Comment by erkinalp** [comment]
> Threads in mx-puppet-slack are simply flattened into reply chains. I think this could be a good compromise as a quick first step? 

Discord threads are channels, which would map to rooms in Matrix. Quite different from Slack-style threads, and threads-as-relations that Matrix currently has.
